core/src/main/scala/kafka/network/SocketServer.scala

```diff
@@ -1107,8 +1107,10 @@ private[kafka] class Processor(
     val header = RequestHeader.parse(buffer)
     if (apiVersionManager.isApiEnabled(header.apiKey, header.apiVersion)) {
       header
-    } else {
+    } else if (header.isApiVersionDeprecated()) {
       throw new InvalidRequestException(s"Received request api key ${header.apiKey} with version ${header.apiVersion} which is not enabled")
+    } else {
+      throw new UnsupportedVersionException(s"Received request api key ${header.apiKey} with version ${header.apiVersion} which is not supported")
```

Comment:
Pardon me, this changes the returned exception from `INVALID_REQUEST` to `UNSUPPORTED_VERSION` as `isApiVersionDeprecated` always returns `false` for now. Is it expected?

Reply:
No, I added a commit at the end that changed this incorrectly:

https://github.com/apache/kafka/commit/b381c07a595ad616aca2579fb34660f48ea60333

Thanks for catching that.

Reply:
Submitted the fix: https://github.com/apache/kafka/pull/18340